File: rb/test/get_polling_database_entity_test.rb
# GetPollingDatabase entity test

require "minitest/autorun"
require "json"
require_relative "../WahlumfragenDatenbank_sdk"
require_relative "runner"

class GetPollingDatabaseEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WahlumfragenDatenbankSDK.test(nil, nil)
    ent = testsdk.GetPollingDatabase(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = get_polling_database_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "get_polling_database." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    get_polling_database_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.get_polling_database")))
    get_polling_database_ref01_data = nil
    if get_polling_database_ref01_data_raw.length > 0
      get_polling_database_ref01_data = Helpers.to_map(get_polling_database_ref01_data_raw[0][1])
    end

    # LIST
    get_polling_database_ref01_ent = client.GetPollingDatabase(nil)
    get_polling_database_ref01_match = {}

    get_polling_database_ref01_list_result, err = get_polling_database_ref01_ent.list(get_polling_database_ref01_match, nil)
    assert_nil err
    assert get_polling_database_ref01_list_result.is_a?(Array)

  end
end

def get_polling_database_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "get_polling_database", "GetPollingDatabaseTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WahlumfragenDatenbankSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["get_polling_database01", "get_polling_database02", "get_polling_database03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID" => idmap,
    "WAHLUMFRAGENDATENBANK_TEST_LIVE" => "FALSE",
    "WAHLUMFRAGENDATENBANK_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WAHLUMFRAGENDATENBANK_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = WahlumfragenDatenbankSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WAHLUMFRAGENDATENBANK_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WAHLUMFRAGENDATENBANK_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/get_polling_database_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/wahlumfragen-datenbank-sdk/go"
	"github.com/voxgig-sdk/wahlumfragen-datenbank-sdk/go/core"

	vs "github.com/voxgig-sdk/wahlumfragen-datenbank-sdk/go/utility/struct"
)

func TestGetPollingDatabaseEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.GetPollingDatabase(nil)
		if ent == nil {
			t.Fatal("expected non-nil GetPollingDatabaseEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := get_polling_databaseBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "get_polling_database." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		getPollingDatabaseRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.get_polling_database", setup.data)))
		var getPollingDatabaseRef01Data map[string]any
		if len(getPollingDatabaseRef01DataRaw) > 0 {
			getPollingDatabaseRef01Data = core.ToMapAny(getPollingDatabaseRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = getPollingDatabaseRef01Data

		// LIST
		getPollingDatabaseRef01Ent := client.GetPollingDatabase(nil)
		getPollingDatabaseRef01Match := map[string]any{}

		getPollingDatabaseRef01ListResult, err := getPollingDatabaseRef01Ent.List(getPollingDatabaseRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, getPollingDatabaseRef01ListOk := getPollingDatabaseRef01ListResult.([]any)
		if !getPollingDatabaseRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", getPollingDatabaseRef01ListResult)
		}

	})
}

func get_polling_databaseBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "get_polling_database", "GetPollingDatabaseTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read get_polling_database test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse get_polling_database test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"get_polling_database01", "get_polling_database02", "get_polling_database03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID": idmap,
		"WAHLUMFRAGENDATENBANK_TEST_LIVE":      "FALSE",
		"WAHLUMFRAGENDATENBANK_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["WAHLUMFRAGENDATENBANK_TEST_GET_POLLING_DATABASE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WAHLUMFRAGENDATENBANK_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewWahlumfragenDatenbankSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WAHLUMFRAGENDATENBANK_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WAHLUMFRAGENDATENBANK_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
